src/metabase/api/query_metadata.clj

```diff
@@ -3,135 +3,142 @@
+  (-> (reduce (fn [m src]
+                (if-let [card-id (lib.util/legacy-string-table-id->card-id src)]
+                  (update m :cards conj! card-id)
+                  (update m :tables conj! src)))
+              {:cards  (transient #{})
+               :tables (transient #{})}
+              source-ids)
+      (update-vals persistent!)))
+
+(defn- query->template-tag-field-ids [query]
+  (when-let [template-tags (some-> query :native :template-tags vals seq)]
+    (for [{tag-type :type, [dim-tag id _opts] :dimension} template-tags
+          :when (and (= tag-type :dimension)
+                     (= dim-tag :field)
+                     (integer? id))]
+      id)))
 
-(defn- metadata-for-table-dependents
-  [table-dependents]
-  (let [{:keys [card-ids table-ids]} (partition-table-ids (map :id table-dependents))
-        card-metadatas (api.table/batch-fetch-card-query-metadatas card-ids)
-        table-metadatas (api.table/batch-fetch-table-query-metadatas table-ids)
-        all-metadatas (-> []
-                          (into (sort-by :id table-metadatas))
-                          (into (sort-by :id card-metadatas)))]
-    (when (< (count all-metadatas) (count table-dependents))
-      (log/warn "Some (possibly virtual) tables are not readable by the current user"))
-    all-metadatas))
+(defn- batch-fetch-query-metadata*
+  "Fetch dependent metadata for ad-hoc queries."
+  [queries]
+  (let [source-ids                (into #{} (mapcat #(lib.util/collect-source-tables (:query %)))
+                                        queries)
+        {source-table-ids :tables
+         source-card-ids  :cards} (split-tables-and-legacy-card-refs source-ids)
+        source-tables             (concat (api.table/batch-fetch-table-query-metadatas source-table-ids)
+                                          (api.table/batch-fetch-card-query-metadatas source-card-ids))
+        fk-target-field-ids       (into #{} (comp (mapcat :fields)
+                                                  (keep :fk_target_field_id))
+                                        source-tables)
+        fk-target-table-ids       (into #{} (remove source-table-ids)
+                                        (field-ids->table-ids fk-target-field-ids))
+        fk-target-tables          (api.table/batch-fetch-table-query-metadatas fk-target-table-ids)
+        tables                    (concat source-tables fk-target-tables)
+        template-tag-field-ids    (into #{} (mapcat query->template-tag-field-ids) queries)
+        query-database-ids        (into #{} (keep :database) queries)
+        database-ids              (into query-database-ids
+                                        (keep :db_id)
+                                        tables)]
+    {;; TODO: This is naive and issues multiple queries currently. That's probably okay for most dashboards,
+     ;; since they tend to query only a handful of databases at most.
+     :databases (->> (for [id database-ids]
```

Comment:
@bshepherdson will it throw an exception if the database is not accessible? I think we should add a simple `get-databases` like `get-cards`.